File: loopx/control_plane/turn_driver/delivery_continuity.ts
import type { JsonObject } from "../effect_program.ts";
import {
  optionalNonEmptyString,
  requireBoolean,
  requireJsonObject,
  requireNonEmptyString,
  requireStringLiteral,
} from "../runtime_decode.ts";
import {
  decodeOptionalMaterialDeliveryOutcome,
  type MaterialDeliveryOutcome,
} from "../work_items/delivery_outcome.ts";

export const DELIVERY_CONTINUITY_RESULT_SCHEMA =
  "loopx_delivery_continuity_result_v0";
export const DELIVERY_BOUNDARY_RESULT_SCHEMA =
  "loopx_delivery_boundary_result_v0";
export const DELIVERY_ROUTING_REQUEST_SCHEMA =
  "loopx_delivery_routing_request_v0";
export const DELIVERY_ROUTING_RESULT_SCHEMA =
  "loopx_delivery_routing_result_v0";

export const DELIVERY_BOUNDARIES = [
  "in_flight_continuation",
  "semantic_closeout",
] as const;

export const DELIVERY_CONTINUITY_PREEMPTIONS = [
  "heartbeat_receipt",
  "blocking_work_lane",
  "autonomous_replan",
  "control_repair",
  "delivery_not_allowed",
] as const;

export type DeliveryBoundary = (typeof DELIVERY_BOUNDARIES)[number];
export type DeliveryContinuityPreemption =
  (typeof DELIVERY_CONTINUITY_PREEMPTIONS)[number];

type TodoStatus = "open" | "done" | "blocked" | "deferred";
const TODO_STATUSES = ["open", "done", "blocked", "deferred"] as const;

export interface DeliveryContinuityTodo {
  todo_id: string;
  status: TodoStatus;
  task_class: string;
  claimed_by: string | null;
  actionable: boolean;
  capability_ready: boolean;
}

interface DeliveryContinuityFacts {
  agent_id: string;
  previous_todo_id: string | null;
  previous_delivery_outcome: MaterialDeliveryOutcome | null;
  current_todo: DeliveryContinuityTodo | null;
  preemptions: readonly DeliveryContinuityPreemption[];
}

export type DeliveryContinuityDecision =
  | "resume_in_flight"
  | "release_for_reselection"
  | "preempt";

export type DeliveryContinuityReason =
  | DeliveryContinuityPreemption
  | "no_previous_todo"
  | "previous_delivery_not_progress"
  | "current_todo_missing"
  | "todo_identity_changed"
  | "todo_not_open"
  | "todo_not_advancement"
  | "todo_not_actionable"
  | "todo_capability_blocked"
  | "todo_claimed_by_other_agent"
  | "same_open_todo_after_progress";

export type DeliveryContinuityResult = JsonObject & {
  schema_version: typeof DELIVERY_CONTINUITY_RESULT_SCHEMA;
  decision: DeliveryContinuityDecision;
  reason: DeliveryContinuityReason;
  todo_id: string | null;
  delivery_boundary: DeliveryBoundary;
};

interface DeliveryBoundaryFacts {
  agent_id: string;
  current_todo: DeliveryContinuityTodo | null;
  preemptions: readonly DeliveryContinuityPreemption[];
}

export type DeliveryBoundaryReason =
  | DeliveryContinuityPreemption
  | "no_selected_todo"
  | "todo_not_open"
  | "todo_not_advancement"
  | "todo_not_actionable"
  | "todo_capability_blocked"
  | "todo_claimed_by_other_agent"
  | "open_advancement_todo";

export type DeliveryBoundaryResult = JsonObject & {
  schema_version: typeof DELIVERY_BOUNDARY_RESULT_SCHEMA;
  delivery_boundary: DeliveryBoundary;
  reason: DeliveryBoundaryReason;
  todo_id: string | null;
};

export type DeliveryRoutingSelection = "continuity" | "fallback" | "none";

interface DeliveryRoutingRequest {
  schema_version: typeof DELIVERY_ROUTING_REQUEST_SCHEMA;
  agent_id: string;
  previous_todo_id: string | null;
  previous_delivery_outcome: MaterialDeliveryOutcome | null;
  continuity_todo: DeliveryContinuityTodo | null;
  fallback_todo: DeliveryContinuityTodo | null;
  preemptions: readonly DeliveryContinuityPreemption[];
}

export type DeliveryRoutingResult = JsonObject & {
  schema_version: typeof DELIVERY_ROUTING_RESULT_SCHEMA;
  selection: DeliveryRoutingSelection;
  continuity: DeliveryContinuityResult | null;
  boundary: DeliveryBoundaryResult | null;
};

function todoStatus(value: unknown, label: string): TodoStatus {
  return requireStringLiteral(
    value,
    TODO_STATUSES,
    label,
    `${label} is unsupported`,
  );
}

function preemptions(value: unknown): DeliveryContinuityPreemption[] {
  if (!Array.isArray(value)) {
    throw new Error("preemptions must be an array");
  }
  const result: DeliveryContinuityPreemption[] = [];
  for (const item of value) {
    if (
      typeof item !== "string" ||
      !DELIVERY_CONTINUITY_PREEMPTIONS.some((candidate) => candidate === item)
    ) {
      throw new Error("preemptions contains an unsupported reason");
    }
    const reason = item as DeliveryContinuityPreemption;
    if (!result.includes(reason)) result.push(reason);
  }
  return result;
}

function currentTodo(
  value: unknown,
  label = "current_todo",
): DeliveryContinuityTodo | null {
  if (value === null || value === undefined) return null;
  const todo = requireJsonObject(value, label);
  return {
    todo_id: requireNonEmptyString(todo.todo_id, `${label}.todo_id`),
    status: todoStatus(todo.status, `${label}.status`),
    task_class: requireNonEmptyString(todo.task_class, `${label}.task_class`),
    claimed_by: optionalNonEmptyString(todo.claimed_by, `${label}.claimed_by`),
    actionable: requireBoolean(todo.actionable, `${label}.actionable`),
    capability_ready: requireBoolean(
      todo.capability_ready,
      `${label}.capability_ready`,
    ),
  };
}

function result(
  decision: DeliveryContinuityDecision,
  reason: DeliveryContinuityReason,
  todoId: string | null,
): DeliveryContinuityResult {
  return {
    schema_version: DELIVERY_CONTINUITY_RESULT_SCHEMA,
    decision,
    reason,
    todo_id: todoId,
    delivery_boundary: decision === "resume_in_flight"
      ? "in_flight_continuation"
      : "semantic_closeout",
  };
}

function resolveDeliveryContinuity(
  request: DeliveryContinuityFacts,
): DeliveryContinuityResult {
  const todoId = request.previous_todo_id;
  const preemption = request.preemptions[0];
  if (preemption !== undefined) {
    return result("preempt", preemption, todoId);
  }
  if (todoId === null) {
    return result("release_for_reselection", "no_previous_todo", null);
  }
  if (request.previous_delivery_outcome !== "outcome_progress") {
    return result(
      "release_for_reselection",
      "previous_delivery_not_progress",
      todoId,
    );
  }
  const todo = request.current_todo;
  if (todo === null) {
    return result("release_for_reselection", "current_todo_missing", todoId);
  }
  if (todo.todo_id !== todoId) {
    return result("release_for_reselection", "todo_identity_changed", todoId);
  }
  if (todo.status !== "open") {
    return result("release_for_reselection", "todo_not_open", todoId);
  }
  if (todo.task_class !== "advancement_task") {
    return result("release_for_reselection", "todo_not_advancement", todoId);
  }
  if (!todo.actionable) {
    return result("release_for_reselection", "todo_not_actionable", todoId);
  }
  if (!todo.capability_ready) {
    return result("release_for_reselection", "todo_capability_blocked", todoId);
  }
  if (todo.claimed_by !== null && todo.claimed_by !== request.agent_id) {
    return result(
      "release_for_reselection",
      "todo_claimed_by_other_agent",
      todoId,
    );
  }
  return result(
    "resume_in_flight",
    "same_open_todo_after_progress",
    todoId,
  );
}

function boundaryResult(
  deliveryBoundary: DeliveryBoundary,
  reason: DeliveryBoundaryReason,
  todoId: string | null,
): DeliveryBoundaryResult {
  return {
    schema_version: DELIVERY_BOUNDARY_RESULT_SCHEMA,
    delivery_boundary: deliveryBoundary,
    reason,
    todo_id: todoId,
  };
}

function resolveDeliveryBoundary(
  request: DeliveryBoundaryFacts,
): DeliveryBoundaryResult {
  const todo = request.current_todo;
  const todoId = todo?.todo_id ?? null;
  const preemption = request.preemptions[0];
  if (preemption !== undefined) {
    return boundaryResult("semantic_closeout", preemption, todoId);
  }
  if (todo === null) {
    return boundaryResult("semantic_closeout", "no_selected_todo", null);
  }
  if (todo.status !== "open") {
    return boundaryResult("semantic_closeout", "todo_not_open", todoId);
  }
  if (todo.task_class !== "advancement_task") {
    return boundaryResult("semantic_closeout", "todo_not_advancement", todoId);
  }
  if (!todo.actionable) {
    return boundaryResult("semantic_closeout", "todo_not_actionable", todoId);
  }
  if (!todo.capability_ready) {
    return boundaryResult("semantic_closeout", "todo_capability_blocked", todoId);
  }
  if (todo.claimed_by !== null && todo.claimed_by !== request.agent_id) {
    return boundaryResult(
      "semantic_closeout",
      "todo_claimed_by_other_agent",
      todoId,
    );
  }
  return boundaryResult(
    "in_flight_continuation",
    "open_advancement_todo",
    todoId,
  );
}

function decodeDeliveryRoutingRequest(value: unknown): DeliveryRoutingRequest {
  const request = requireJsonObject(value, "turn.delivery_route params");
  if (request.schema_version !== DELIVERY_ROUTING_REQUEST_SCHEMA) {
    throw new Error("Delivery routing request schema mismatch");
  }
  return {
    schema_version: DELIVERY_ROUTING_REQUEST_SCHEMA,
    agent_id: requireNonEmptyString(request.agent_id, "agent_id"),
    previous_todo_id: optionalNonEmptyString(
      request.previous_todo_id,
      "previous_todo_id",
    ),
    previous_delivery_outcome: decodeOptionalMaterialDeliveryOutcome(
      request.previous_delivery_outcome,
      "previous_delivery_outcome",
    ),
    continuity_todo: currentTodo(request.continuity_todo, "continuity_todo"),
    fallback_todo: currentTodo(request.fallback_todo, "fallback_todo"),
    preemptions: preemptions(request.preemptions),
  };
}

/** Resolve one quota delivery route behind one runtime request/response. */
export function evaluateDeliveryRoute(value: unknown): DeliveryRoutingResult {
  const request = decodeDeliveryRoutingRequest(value);
  const continuity = request.previous_todo_id === null
    ? null
    : resolveDeliveryContinuity({
      agent_id: request.agent_id,
      previous_todo_id: request.previous_todo_id,
      previous_delivery_outcome: request.previous_delivery_outcome,
      current_todo: request.continuity_todo,
      preemptions: request.preemptions,
    });
  const selection: DeliveryRoutingSelection =
    continuity?.decision === "resume_in_flight"
      ? "continuity"
      : request.fallback_todo === null
      ? "none"
      : "fallback";
  const selectedTodo = selection === "continuity"
    ? request.continuity_todo
    : selection === "fallback"
    ? request.fallback_todo
    : null;
  const boundary = selectedTodo === null
    ? null
    : resolveDeliveryBoundary({
      agent_id: request.agent_id,
      current_todo: selectedTodo,
      preemptions: request.preemptions,
    });
  return {
    schema_version: DELIVERY_ROUTING_RESULT_SCHEMA,
    selection,
    continuity,
    boundary,
  };
}

File: loopx/control_plane/turn_driver/delivery_continuity.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from ..effect_runtime import EffectRuntimeRejected, effect_runtime_result
from ..todos.contract import (
    TODO_STATUS_OPEN,
    normalize_todo_claimed_by,
    normalize_todo_id,
    normalize_todo_status,
)
from ..todos.projection import todo_item_task_class


DELIVERY_CONTINUITY_REQUEST_SCHEMA = "loopx_delivery_continuity_request_v0"
DELIVERY_CONTINUITY_RESULT_SCHEMA = "loopx_delivery_continuity_result_v0"
DELIVERY_BOUNDARY_REQUEST_SCHEMA = "loopx_delivery_boundary_request_v0"
DELIVERY_BOUNDARY_RESULT_SCHEMA = "loopx_delivery_boundary_result_v0"
DELIVERY_BOUNDARY_IN_FLIGHT = "in_flight_continuation"
DELIVERY_BOUNDARY_SEMANTIC_CLOSEOUT = "semantic_closeout"
DELIVERY_BOUNDARY_VALUES = (
    DELIVERY_BOUNDARY_IN_FLIGHT,
    DELIVERY_BOUNDARY_SEMANTIC_CLOSEOUT,
)
DELIVERY_CONTINUITY_DECISIONS = {
    "resume_in_flight",
    "release_for_reselection",
    "preempt",
}
DELIVERY_CONTINUITY_PREEMPTIONS = {
    "heartbeat_receipt",
    "blocking_work_lane",
    "autonomous_replan",
    "control_repair",
    "delivery_not_allowed",
}
DELIVERY_BOUNDARY_REASONS = {
    *DELIVERY_CONTINUITY_PREEMPTIONS,
    "no_selected_todo",
    "todo_not_open",
    "todo_not_advancement",
    "todo_not_actionable",
    "todo_capability_blocked",
    "todo_claimed_by_other_agent",
    "open_advancement_todo",
}


def normalize_delivery_boundary(value: Any) -> str:
    boundary = str(value or DELIVERY_BOUNDARY_SEMANTIC_CLOSEOUT).strip()
    if boundary not in DELIVERY_BOUNDARY_VALUES:
        raise ValueError(
            "delivery_boundary must be one of: " + ", ".join(DELIVERY_BOUNDARY_VALUES)
        )
    return boundary


def _delivery_todo_payload(
    current_todo: Mapping[str, Any] | None,
    *,
    actionable: bool,
    capability_ready: bool,
) -> dict[str, Any] | None:
    if current_todo is None:
        return None
    todo_id = normalize_todo_id(current_todo.get("todo_id"))
    if not todo_id:
        raise ValueError("delivery continuity current Todo requires a typed todo_id")
    return {
        "todo_id": todo_id,
        "status": normalize_todo_status(current_todo.get("status")) or TODO_STATUS_OPEN,
        "task_class": todo_item_task_class(dict(current_todo)),
        "claimed_by": normalize_todo_claimed_by(current_todo.get("claimed_by")),
        "actionable": bool(actionable),
        "capability_ready": bool(capability_ready),
    }


def _normalize_preemptions(preemptions: Sequence[str]) -> list[str]:
    normalized = list(dict.fromkeys(str(item).strip() for item in preemptions))
    if any(item not in DELIVERY_CONTINUITY_PREEMPTIONS for item in normalized):
        raise ValueError("delivery continuity received an unsupported preemption")
    return normalized


def evaluate_delivery_continuity(
    *,
    agent_id: str,
    previous_todo_id: str | None,
    previous_delivery_outcome: str | None,
    current_todo: Mapping[str, Any] | None,
    actionable: bool,
    capability_ready: bool,
    preemptions: Sequence[str] = (),
) -> dict[str, Any]:
    """Ask the TypeScript Turn owner whether one open Todo remains in flight."""

    safe_agent_id = normalize_todo_claimed_by(agent_id)
    if not safe_agent_id:
        raise ValueError("delivery continuity requires a valid agent_id")
    normalized_preemptions = _normalize_preemptions(preemptions)
    todo_payload = _delivery_todo_payload(
        current_todo,
        actionable=actionable,
        capability_ready=capability_ready,
    )
    try:
        result = effect_runtime_result(
            "turn.delivery_continuity.evaluate",
            {
                "schema_version": DELIVERY_CONTINUITY_REQUEST_SCHEMA,
                "agent_id": safe_agent_id,
                "previous_todo_id": normalize_todo_id(previous_todo_id),
                "previous_delivery_outcome": (
                    str(previous_delivery_outcome or "").strip() or None
                ),
                "current_todo": todo_payload,
                "preemptions": normalized_preemptions,
            },
        )
    except EffectRuntimeRejected as exc:
        raise ValueError(str(exc)) from None
    if not isinstance(result, Mapping):
        raise RuntimeError("TypeScript delivery continuity result must be an object")
    if (
        result.get("schema_version") != DELIVERY_CONTINUITY_RESULT_SCHEMA
        or result.get("decision") not in DELIVERY_CONTINUITY_DECISIONS
        or not isinstance(result.get("reason"), str)
        or result.get("delivery_boundary") not in DELIVERY_BOUNDARY_VALUES
        or not (result.get("todo_id") is None or isinstance(result.get("todo_id"), str))
    ):
        raise RuntimeError("TypeScript delivery continuity result shape mismatch")
    return dict(result)


def evaluate_delivery_boundary(
    *,
    agent_id: str,
    current_todo: Mapping[str, Any] | None,
    actionable: bool,
    capability_ready: bool,
    preemptions: Sequence[str] = (),
) -> dict[str, Any]:
    """Ask the TypeScript Turn owner how this selected Todo should settle."""

    safe_agent_id = normalize_todo_claimed_by(agent_id)
    if not safe_agent_id:
        raise ValueError("delivery boundary requires a valid agent_id")
    try:
        result = effect_runtime_result(
            "turn.delivery_boundary.evaluate",
            {
                "schema_version": DELIVERY_BOUNDARY_REQUEST_SCHEMA,
                "agent_id": safe_agent_id,
                "current_todo": _delivery_todo_payload(
                    current_todo,
                    actionable=actionable,
                    capability_ready=capability_ready,
                ),
                "preemptions": _normalize_preemptions(preemptions),
            },
        )
    except EffectRuntimeRejected as exc:
        raise ValueError(str(exc)) from None
    if not isinstance(result, Mapping):
        raise RuntimeError("TypeScript delivery boundary result must be an object")
    if (
        result.get("schema_version") != DELIVERY_BOUNDARY_RESULT_SCHEMA
        or result.get("delivery_boundary") not in DELIVERY_BOUNDARY_VALUES
        or result.get("reason") not in DELIVERY_BOUNDARY_REASONS
        or not (result.get("todo_id") is None or isinstance(result.get("todo_id"), str))
    ):
        raise RuntimeError("TypeScript delivery boundary result shape mismatch")
    return dict(result)
